presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -689,26 +710,59 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
         Object left = ExpressionInterpreter.expressionOptimizer(comparison.getLeft(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
         Object right = ExpressionInterpreter.expressionOptimizer(comparison.getRight(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right)));
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
+            return Optional.empty();
+        }
+
+        if (left instanceof Expression) {
+            return toNormalizedSimpleComparison(metadata, expressionTypes, comparison.getLeft(), comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right));
+        }
+        else {
+            return toNormalizedSimpleComparison(metadata, expressionTypes, comparison.getRight(), comparison.getType().flip(), new NullableValue(expressionTypes.get(comparison.getLeft()), left));
         }
-        if (right instanceof SymbolReference && !(left instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) right, comparison.getType().flip(), new NullableValue(expressionTypes.get(comparison.getLeft()), left)));
+    }
+
+    private static Optional<NormalizedSimpleComparison> toNormalizedSimpleComparison(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression nonLiteral, ComparisonExpressionType comparisonType, NullableValue literal)
+    {
+        boolean nonCoercionCastRemoved = false;
+        while (true) {
+            if (nonLiteral instanceof SymbolReference) {
+                return Optional.of(new NormalizedSimpleComparison((SymbolReference) nonLiteral, comparisonType, literal, nonCoercionCastRemoved));
+            }
+            else if (nonLiteral instanceof Cast) {
+                if (!isCoercion(metadata, expressionTypes, (Cast) nonLiteral)) {
```

Comment:
So I don't think we should be stripping out casts like this without considering their effect on the comparison at each step. Casts can connect arbitrary types and have arbitrary effects at each step. For example, if we have an INTEGER symbol that is cast to a TINYINT and then a BIGINT. The single cast from INTEGER to BIGINT will be insufficient.

What we should be doing here is applying the transformations to the literal as we unpeel each layer of casts. In this case I *think* the appropriate set of transformations is to apply the implicit coercion directly if possible, otherwise seek out a saturated floor cast, otherwise bail on this operation.